LiteCore/Database/DatabaseImpl.cc

```diff
@@ -626,11 +626,28 @@ namespace litecore {
     C4Collection* DatabaseImpl::getOrCreateCollection(CollectionSpec spec, bool canCreate) {
         // Is there already a C4Collection object for it in _collections?
         LOCK(_collectionsMutex);
-        if (auto i = _collections.find(spec); i != _collections.end())
-            return i->second;                                         // -> Existing object
-
         // Validate the name (throws if invalid):
         string keyStoreName = collectionNameToKeyStoreName(spec);
+        if (auto i = _collections.find(spec); i != _collections.end()) { 
+            if(!i->second->isValid()) {
+                if(!canCreate) {
+                    if(!_dataFile->keyStoreExists(keyStoreName)) {
+                        // This is an invalid collection, and it has not been recreated elsewhere.
+                        // Furthermore, canCreate is false so don't return a collection object
+                        return nullptr;
+                    } 
+                }
+
+                // If we reach here, either canCreate is true and we are about to recreate
+                // a previously deleted collection, or canCreate is false and the deleted collection
+                // was created by another instance already.  Either way, it is time to remove the
+                // old invalid entry now that it is safe to do so.
+                asInternal(i->second)->close();
```

Comment:
If won't ever resurrect the collection object, is it still needed to. have a _valid flag? Could C4Collection::invalidate simply assign _database to NULL?

Reply:
That might be true, I will think about that.